### PR TITLE
new OSCCAL tuner

### PR DESCRIPTION
**Pull request by nerdralph**
Tested several times with a t85@3V3, with my low-speed USB mouse on the same hub.  No enumeration failures, no mn command failures or errors flashing and erasing the target several times.

### CODE REVIEW

**Comment by cpldcpu** [comment]
Perfect, thanks a lot! Looking at it right now.

File: firmware/macros.inc [hunk 1]
```diff
@@ -0,0 +1,25 @@
+; Ralph Doncaster public domain AVR asm macros github.com/nerdralph
```
Comment (cpldcpu) [inline firmware/macros.inc:1]: Is this needed as an extra-file or could it be folded into the osccal source? I see the benefit of having various helper macros. The issue is just that many such includes accumulate over time and if they are only added for specific features it is not very helpful to separate them out.

**Comment by cpldcpu** [comment]
Nice, I can also commit to pull requests. That was new to me.

**Comment by cpldcpu** [comment]
I updated the readme and the hexfiles now. I would like to tag this as 2.5 release when merged, so we finally have that fixed and can set up a new testing branch.

Ralph, thank again for your awesome contribution!

@ArminJo can you check whether everything is ok for you? If possible, it would be great if you could test the new osccal code also on an ATtiny167. I have no such device available.

File: firmware/macros.inc (hunk 1 above)
Comment (nerdralph) [reply firmware/macros.inc:1]: I considered it both ways.  I already had it in a separate file for other asm projects, and my though was to also use it in usbdrv as well.  There's at least one ugly #if block that can be removed:
https://github.com/nerdralph/micronucleus/blob/master/firmware/usbdrv/usbdrvasm.S#L55

**Comment by nerdralph** [comment]
> 
> 
> Nice, I can also commit to pull requests. That was new to me.
> 
> Wait, this is commiting to nerdralph/master? Interesting that this is possible from a permission perspective. So I earned writing rights to your repository because you sent a commmit? Weird, but convenient in this case.

That came as a surprise to me too, as I didn't explicitly add you on my fork.  I guess it reduces the chance of merge conflicts if/when I re-sync to upstream.

**Comment by nerdralph** [comment]
> 
> Ralph, thank again for your awesome contribution!
> 

And thanks for your feedback.  It's nice to bounce ideas off another asm guru.  There's not many of us around...

File: firmware/macros.inc (hunk 1 above)
Comment (nerdralph) [reply firmware/macros.inc:1]: Although I might be able to get rid of the pending interrupt code altogether, in which case there's little reason to keep the macros in a separate file.

**Comment by cpldcpu** [comment]
At least they are hiding somewhere :) Yeah, the situations where you can really utilize highly optimized and timing critical code are getting scarce.

**Comment by cpldcpu** [comment]
Ok, I'll merge this for now because unreleated changes are beginning to accumulate. This was not supposed to be the test branch :)

@ArminJo waiting for your feedback before tagging the release.

**Comment by ArminJo** [comment]
I will test it this week!

**Comment by Ho-Ro** [comment]
> Tested several times with a t85@3V3, with my low-speed USB mouse on the same hub. No enumeration failures, no mn command failures or errors flashing and erasing the target several times.

I can confirm, the enum errors with t85 (digispark) on my laptop with an USB mouse connected to another port of the same USB bus (i.e. internal hub) didn't show up anymore with default bootloader 2.5. Thx for solving this nasty issue!

**Comment by nerdralph** [comment]
> 
> 
> > Tested several times with a t85@3V3, with my low-speed USB mouse on the same hub. No enumeration failures, no mn command failures or errors flashing and erasing the target several times.
> 
> I can confirm, the enum errors with t85 (digispark) on my laptop with an USB mouse connected to another port of the same USB bus (i.e. internal hub) didn't show up anymore with default bootloader 2.5. Thx for solving this nasty issue!

Good to have additional confirmation.  Thanks.
Although I haven't finished converting all of MN to asm, the current version has been stable in my testing.  I wouldn't mind hearing how others make out with it.
https://github.com/nerdralph/micronucleus-firmware/tree/asm

**Comment by ArminJo** [comment]
I tested with my digispark, digispark-pro and MH-ET LIVE Tiny88 (16.0MHz) board and everything was OK. Saving was 22 bytes which I spent for extending the FAST_EXIT_NO_USB_MS functionality 😀 . Thanks

**Comment by nerdralph** [comment]
The new tuner isn't quite perfect.  With t85_default + SAVE_CALIB 0, I sometimes have to unplug and reconnect in order to get reliable flashing to work.  I'm going to do some testing with one extra frame for calibration.

**Comment by cpldcpu** [comment]
@ArminJo 

Awesome! I will tag the release then.  Let's push further improvements to testing first, unless there are important bug fixes.

**Pull request merged**